### PR TITLE
Add ostree=aboot for signed Android Boot Images

### PR DESCRIPTION
**Pull request by ericcurtin**
Some kernel images are delivered in a signed kernel + cmdline + initramfs + dtb blob. When this is added to the commit server side, only after this do you know what the cmdline is, this creates a recursion issue. To avoid this, in the case where we have ostree=aboot karg set, create a symlink after deploy to the correct ostree target in the rootfs, as the cmdline can't be malleable and secured client-side at the same time.

### CODE REVIEW

**Comment by ericcurtin** [comment]
So this is #2844 reworked, much simpler

File: src/switchroot/ostree-mount-util.h [hunk 1]
```diff
@@ -114,6 +120,46 @@ read_proc_cmdline_key (const char *key)
   return ret;
 }
 
+static inline char *
+get_aboot_root_slot (void)
+{
+  char __attribute__ ((cleanup (free_char))) *slot_suffix
```
Comment (alexlarsson) [inline src/switchroot/ostree-mount-util.h:126]: I think if we start using cleanup attribute in prepare-root we should add a define for this.

Comment (ericcurtin) [reply src/switchroot/ostree-mount-util.h:126]: It's actually recommended in the coding style "Use GCC `__attribute__((cleanup))` wherever possible. ", not sure how many read it, no extra dependencies and is nicer than manual cleanup.

I'll add some sort of define, the real string is long winded.

File: src/switchroot/ostree-mount-util.h [hunk 2]
```diff
@@ -114,6 +120,46 @@ read_proc_cmdline_key (const char *key)
...
+  else if (slot_suffix[1] == 'b')
+    return strdup ("/ostree/root.b");
+
+  errx (EXIT_FAILURE, "androidboot.slot_suffix invalid: %s", slot_suffix);
+
+  return NULL;
+}
+
+static inline char *
+move (char **to_move)
```
Comment (alexlarsson) [inline src/switchroot/ostree-mount-util.h:139]: This is typically called steal_pointer()

File: src/switchroot/ostree-mount-util.h [hunk 3]
```diff
@@ -114,6 +120,46 @@ read_proc_cmdline_key (const char *key)
...
+
+static inline char *
+get_ostree_target (void)
+{
+  char __attribute__ ((cleanup (free_char))) *ostree_cmdline = read_proc_cmdline_key ("ostree");
+
+  if (!ostree_cmdline)
+    errx (EXIT_FAILURE, "No ostree= cmdline");
+
+  if (strcmp (ostree_cmdline, "aboot"))
```
Comment (alexlarsson) [inline src/switchroot/ostree-mount-util.h:155]: Personally i like strcmp calls to explicitly compare with zero. so `== 0` or ` != 0` just because the result really isn't a traditional boolean.

File: src/switchroot/ostree-prepare-root.c [hunk 4]
```diff
@@ -249,7 +246,7 @@ main (int argc, char *argv[])
     }
 
   OstreeComposefsMode composefs_mode = OSTREE_COMPOSEFS_MODE_MAYBE;
-  char *ot_composefs = read_proc_cmdline_key ("ot-composefs");
+  char __attribute__ ((cleanup (free_char))) *ot_composefs = read_proc_cmdline_key ("ot-composefs");
```
Comment (alexlarsson) [inline src/switchroot/ostree-prepare-root.c:249]: This is cool, but ideally split up different fixes in separate commits

File: src/switchroot/ostree-mount-util.h (hunk 2 above)
Comment (ericcurtin) [reply src/switchroot/ostree-mount-util.h:139]: move is a C++ism for the same thing, can change to steal_pointer

Comment (alexlarsson) [reply src/switchroot/ostree-mount-util.h:139]: Well, it is g_steal_pointer() in glib which is used in the rest of ostree.

**Comment by alexlarsson** [comment]
So, where are the symlinks created?

File: src/switchroot/ostree-mount-util.h (hunk 1 above)
Comment (alexlarsson) [reply src/switchroot/ostree-mount-util.h:126]: Yeah, this is used in the rest of ostree but called `g_autofree`. So, maybe just #define it to `autofree`.

**Comment by ericcurtin** [comment]
So at deploy time on the client-side, we shell out to aboot-deploy, this knows where to symlink to and can create the symlink. Doesn't need a change in this repo

**Comment by ericcurtin** [comment]
All review comments addressed

File: src/switchroot/ostree-mount-util.h [hunk 5]
```diff
@@ -40,6 +40,8 @@
 #define _OSTREE_SYSROOT_READONLY_STAMP "/run/ostree-sysroot-ro.stamp"
 #define _OSTREE_COMPOSEFS_ROOT_STAMP "/run/ostree-composefs-root.stamp"
 
+#define autofree_char char __attribute__ ((__cleanup__ (free_char)))
```
Comment (alexlarsson) [inline src/switchroot/ostree-mount-util.h:43]: I don't think you need the _char part here, or in free_char. This can free anything that malloc returned. I would propose just "autofree" define and "free_ptr()".

**Comment by ericcurtin** [comment]
On a slightly related note, I was thinking of replacing https://github.com/ostreedev/ostree/blob/main/src/libostree/ostree-bootloader-aboot.c with a generic c file that calls a systemd unit, I thought of doing something similar and I actually saw a comment that hinted towards that in the code base here somewhere....

Would mean you don't have to write a load of bolierplate C to shell out to something, it would just become a systemd unit file...

File: src/switchroot/ostree-mount-util.h [hunk 6]
```diff
@@ -114,6 +131,34 @@ read_proc_cmdline_key (const char *key)
   return ret;
 }
 
+static inline char *
+get_aboot_root_slot (void)
+{
+  autofree_char *slot_suffix = read_proc_cmdline_key ("androidboot.slot_suffix");
+  if (strcmp (slot_suffix, "_a") == 0)
+    return strdup ("/ostree/root.a");
```
[the remark this answers is not in the document]
Comment (ericcurtin) [reply src/switchroot/ostree-mount-util.h:139]: So we already create links in /ostree:

```
$ ls -ltr /ostree/
total 12
drwxr-xr-x. 3 root root 4096 May 25 17:29 boot.1.1
lrwxrwxrwx. 1 root root    8 May 25 17:29 boot.1 -> boot.1.1
drwxr-xr-x. 7 root root 4096 May 25 17:29 repo
drwxr-xr-x. 3 root root 4096 May 25 17:29 deploy
```

/ostree/deploy/ normally contains osname:

```
ls -ltr /ostree/deploy/
total 4
drwxr-xr-x. 4 root root 4096 May 25 17:29 centos
```

would rather keep in /ostree/ as it's not standardised to store osnames and it's a shorter path.

[the remark this answers is not in the document]
Comment (ericcurtin) [reply src/switchroot/ostree-mount-util.h:139]: ostree expects everything in /ostree/deploy to be an osname (centos, fedora)

File: src/switchroot/ostree-mount-util.h [hunk 7]
```diff
@@ -79,6 +81,21 @@ read_proc_cmdline (void)
...
 }
 
+static inline void
+free_char (char **to_free)
+{
+  free (*to_free);
+}
+
+static inline char *
+steal_pointer_char (char **to_steal)
```
[the remark this answers is not in the document]
Comment (ericcurtin) [reply src/switchroot/ostree-mount-util.h]: Made this generic

File: src/switchroot/ostree-mount-util.h (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexlarsson) [reply src/switchroot/ostree-mount-util.h:141]: makes sense

File: src/switchroot/ostree-prepare-root.c (hunk 4 above)
Comment (ericcurtin) [reply src/switchroot/ostree-prepare-root.c]: They are separate now

**Review by alexlarsson — APPROVED** [review]
LGTM

**Comment by cgwalters** [comment]
I gave this a quick skim and it looks sane.  I do have a vague feeling it might be nicer to try to abstract away from aboot in some way because there's certain to be some alignment as discussed between UKIs and aboot (and potentially other A/B bootloaders) but this is fine.

A procedural note: In this organization we only need one approving reviewer, so in the general case it's not just me that can merge things.  I'm certainly fine if Alex or Dan had merged this too.
